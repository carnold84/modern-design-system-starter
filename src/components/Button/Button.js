import { bool, string } from 'prop-types';
import React from 'react';

import { Wrapper } from './Button.styles';

/**
 * Use `Button` to trigger actions and link to pages / new urls. Use `as` prop to change from the default `button`.
 */
const Button = ({ as, children, isDisabled, isPrimary, type, ...rest }) => {
  return (
    <Wrapper as={as} className={isPrimary ? 'is-primary' : null} disabled={isDisabled} type={type} {...rest}>
      {children}
    </Wrapper>
  );
};

Button.propTypes = {
  as: string,
  isDisabled: bool,
  isPrimary: bool,
  type: string,
};

Button.defaultProps = {
  as: 'button',
  isDisabled: false,
  isPrimary: false,
};

export default Button;
