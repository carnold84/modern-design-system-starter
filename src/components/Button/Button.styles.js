import styled from 'styled-components';

export const Wrapper = styled.button`
  align-items: center;
  background-color: ${props => props.theme.components.button.bgColor};
  border: ${props => props.theme.components.button.borderColor} solid 1px;
  border-radius: 3px;
  color: ${props => props.theme.components.button.color};
  cursor: pointer;
  display: inline-flex;
  font-size: 1rem;
  height: 40px;
  justify-content: center;
  outline: transparent solid 2px;
  padding: 0 15px;
  transition: all 200ms ease-in-out;

  a {
    align-items: center;
    color: ${props => props.theme.components.button.color};
    display: flex;
    height: 100%;
    text-decoration: none;
  }

  &:disabled {
    opacity: 0.75;
    pointer-events: none;
  }

  &:focus {
    background-color: ${props => props.theme.components.button._focus.bgColor};
    border-color: ${props => props.theme.components.button._focus.borderColor};
    color: ${props => props.theme.components.button._focus.color};
    box-shadow: 0 0 0 2px ${props => props.theme.components.button._focus.outlineColor};

    a {
      color: ${props => props.theme.components.button._focus.color};
    }
  }

  &:hover {
    background-color: ${props => props.theme.components.button._hover.bgColor};
    border-color: ${props => props.theme.components.button._hover.borderColor};
    color: ${props => props.theme.components.button._hover.color};

    a {
      color: ${props => props.theme.components.button._hover.color};
    }
  }

  &.is-primary {
    background-color: ${props => props.theme.components.button.primary.bgColor};
    border-color: ${props => props.theme.components.button.primary.borderColor};
    color: ${props => props.theme.components.button.primary.color};

    a {
      color: ${props => props.theme.components.button.primary.color};
    }

    &:focus {
      background-color: ${props => props.theme.components.button.primary._focus.bgColor};
      border-color: ${props => props.theme.components.button.primary._focus.borderColor};
      color: ${props => props.theme.components.button.primary._focus.color};
      box-shadow: 0 0 0 2px ${props => props.theme.components.button.primary._focus.outlineColor};

      a {
        color: ${props => props.theme.components.button.primary._focus.color};
      }
    }

    &:hover {
      background-color: ${props => props.theme.components.button.primary._hover.bgColor};
      border-color: ${props => props.theme.components.button.primary._hover.borderColor};
      color: ${props => props.theme.components.button.primary._hover.color};

      a {
        color: ${props => props.theme.components.button.primary._hover.color};
      }
    }
  }
`;

Wrapper.defaultProps = {
  theme: {
    components: {
      button: {
        _focus: {
          bgColor: '#ffffff',
          borderColor: '#cecece',
          color: '#222222',
          outlineColor: '#cecece',
        },
        _hover: {
          bgColor: '#f3f3f3',
          borderColor: '#cecece',
          color: '#222222',
        },
        bgColor: '#ffffff',
        borderColor: '#dbdbdb',
        color: '#666666',
        primary: {
          _focus: {
            bgColor: '#7f226b',
            borderColor: '#973381',
            color: '#ffffff',
            outlineColor: '#973381',
          },
          _hover: {
            bgColor: '#973381',
            borderColor: '#7f226b',
            color: '#ffffff',
          },
          bgColor: '#7f226b',
          borderColor: '#7f226b',
          color: '#ffffff',
        },
      },
    },
  },
};
