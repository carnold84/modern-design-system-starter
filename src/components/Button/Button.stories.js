import React from 'react';

import Button from './Button';

export default {
  component: Button,
  title: 'Components|Button',
};

export const allButtons = () => {
  return (
    <div>
      <Button onClick={() => alert('Click!')} style={{ margin: '0 10px 0 0' }}>
        Button
      </Button>

      <Button isPrimary={true} onClick={() => alert('Click!')} style={{ margin: '0 10px 0 0' }}>
        Primary Button
      </Button>

      <Button isDisabled={true} onClick={() => alert('Click!')}>
        Button
      </Button>
    </div>
  );
};
