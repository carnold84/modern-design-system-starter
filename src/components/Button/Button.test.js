import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

test('Renders Button', () => {
  const component = renderer.create(<Button>Test</Button>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders Primary Button', () => {
  const component = renderer.create(<Button isPrimary={true}>Test</Button>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders Disabled Button', () => {
  const component = renderer.create(<Button isDisabled={true}>Test</Button>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
