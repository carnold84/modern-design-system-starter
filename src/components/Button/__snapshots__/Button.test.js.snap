// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Button 1`] = `
<button
  className="sc-bdVaJa cutXAD"
  disabled={false}
>
  Test
</button>
`;

exports[`Renders Disabled Button 1`] = `
<button
  className="sc-bdVaJa cutXAD"
  disabled={true}
>
  Test
</button>
`;

exports[`Renders Primary Button 1`] = `
<button
  className="is-primary sc-bdVaJa cutXAD"
  disabled={false}
>
  Test
</button>
`;
